File: python/Largest_Magic_Square_1895.py
import numpy as np

class Solution:
    def largestMagicSquare(self, grid: List[List[int]]) -> int:
        horSum = np.array(grid)
        verSum = np.array(grid)

        for j in range(1, len(horSum[0])):
            for i in range(len(horSum)):
                horSum[i, j] += horSum[i, j - 1]
        
        for i in range(1, len(verSum)):
            for j in range(len(verSum[0])):
                verSum[i, j] += verSum[i - 1, j]

        curMax = 1
        
        for sqrSize in range(min(len(grid), len(grid[0])), 1, -1):
            for i in range(len(grid) - sqrSize + 1):
                for j in range(len(grid[0]) - sqrSize + 1):
                    matchSum = horSum[i, j + sqrSize - 1] - (horSum[i, j - 1] if j > 0 else 0)
                    leftDiag, rightDiag = 0, 0
                    allSumsMatched = True
                    for k in range(sqrSize):
                        horMatch = horSum[i + k, j + sqrSize - 1] - (horSum[i + k, j - 1] if j > 0 else 0)
                        verMatch = verSum[i + sqrSize - 1, j + k] - (verSum[i - 1, j + k] if i > 0 else 0)
                        if horMatch != matchSum or verMatch != matchSum:
                            allSumsMatched = False
                            break
                        leftDiag += grid[i + k][j + k]
                        rightDiag += grid[i + k][j + sqrSize - 1 - k]
                    if not allSumsMatched:
                        continue
                    if leftDiag == rightDiag and leftDiag == matchSum:
                        curMax = max(curMax, sqrSize)
        
        return curMax

File: cpp/Largest_Magic_Square_1895.cpp
class Solution {
public:

    int largestMagicSquare(vector<vector<int>>& grid) {

        vector <vector <int>> horSum = grid, verSum = grid;
        for (int j = 1; j < horSum[0].size(); j++) {
            for (int i = 0; i < horSum.size(); i++) {
                horSum[i][j] += horSum[i][j - 1];
            }
        }
        for (int i = 1; i < verSum.size(); i++) {
            for (int j = 0; j < verSum[i].size(); j++) {
                verSum[i][j] += verSum[i - 1][j];
            }
        }

        int curMax = 1;

        for (int sqrSize = min(grid.size(), grid[0].size()); sqrSize > 1; sqrSize--) {
            for (int i = 0; i + sqrSize - 1 < grid.size(); i++) {
                for (int j = 0; j + sqrSize - 1 < grid[i].size(); j++) {
                    int matchSum = (horSum[i][j + sqrSize - 1] - (j > 0 ? horSum[i][j - 1] : 0));
                    int leftDiag = 0, rightDiag = 0;
                    bool allSumsMatched = true;
                    for (int k = 0; k < sqrSize; k++) {
                        int horMatch = horSum[i + k][j + sqrSize - 1] - (j > 0 ? horSum[i + k][j - 1] : 0);
                        int verMatch = verSum[i + sqrSize - 1][j + k] - (i > 0 ? verSum[i - 1][j + k] : 0);
                        if (horMatch != matchSum || verMatch != matchSum) {
                            allSumsMatched = false;
                            break;
                        }
                        leftDiag += grid[i + k][j + k];
                        rightDiag += grid[i + k][j + sqrSize - 1 - k];
                    }
                    if (leftDiag != rightDiag || leftDiag != matchSum) continue;
                    if (allSumsMatched) curMax = max(curMax, sqrSize);
                }
            }
        }

        return curMax;

    }
};
